src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java

```diff
@@ -0,0 +1,118 @@
+ * between two character sequences.
+ *
+ * <p>
+ * The Sørensen–Dice coefficient is a statistic used for comparing the
+ * similarity of two samples. It was independently developed by the botanists
+ * Thorvald Sørensen and Lee Raymond Dice, who published in 1948 and 1945
+ * respectively. The index is known by several other names, especially
+ * Sørensen–Dice index, Sørensen index and Dice's coefficient. Other variations
+ * include the "similarity coefficient" or "index", such as Dice similarity
+ * coefficient (DSC).
+ * </p>
+ *
+ * <p>
+ * This implementation is based on the Sørensen–Dice similarity algorithm from
+ * <a href=
+ * "http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient">
+ * http://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient</a>.
+ *
+ *
+ * </p>
+ *
+ * @since 1.7
+ */
+public class SorensenDicesSimilarity implements SimilarityScore<Double> {
+
+    /**
+     *
+     * <pre>
+     * similarity.apply("", "")                     = 1.0
+     * similarity.apply("foo", "foo")               = 1.0
+     * similarity.apply("foo", "foo ")              = 0.8
+     * similarity.apply("foo", "foo ")              = 0.66
+     * similarity.apply("foo", " foo ")             = 0.66
+     * similarity.apply("foo", " foo")              = 0.66
+     * similarity.apply("", "a")                    = 0.0
+     * similarity.apply("aaapppp", "")              = 0.0
+     * similarity.apply("frog", "fog")              = 0.4
+     * similarity.apply("fly", "ant")               = 0.0
+     * similarity.apply("elephant", "hippo")        = 0.0
+     * similarity.apply("hippo", "elephant")        = 0.0
+     * similarity.apply("hippo", "zzzzzzzz")        = 0.0
+     * similarity.apply("hello", "hallo")           = 0.5
+     * similarity.apply("ABC Corporation", "ABC Corp") = 0.7
+     * similarity.apply("D N H Enterprises Inc", "D &amp; H Enterprises, Inc.") = 0.74
+     * similarity.apply("My Gym Children's Fitness Center", "My Gym. Childrens Fitness") = 0.81
+     * similarity.apply("PENNSYLVANIA", "PENNCISYLVNIA") = 0.69
+     * </pre>
+     *
+     * @param left  the first CharSequence, must not be null
+     * @param right the second CharSequence, must not be null
+     * @return result similarity
+     * @throws IllegalArgumentException if either CharSequence input is {@code null}
+     */
+    @Override
+    public Double apply(final CharSequence left, final CharSequence right) {
+
+        if (left == null || right == null) {
+            throw new IllegalArgumentException("CharSequences must not be null");
+        }
+
+        if (left.equals(right)) {
+            return 1d;
+        }
+
+        if (left.length() == 0 || right.length() == 0) {
+            return 0d;
+        }
+
+        Set<String> nLeft = createBigrams(left);
+        Set<String> nRight = createBigrams(right);
+
+        final int total = nLeft.size() + nRight.size();
+        nLeft.retainAll(nRight);
```

Comment:
It may not be necessary to call `retainAll` I think? We want to `intersection` being the number of elements common to both vectors/sets/sides/etc.

Indeed, calling `retainAll` does the trick here. But I believe `createBigrams` is returning `HashSet`, which is backed - as far as I can tell - by a `HashMap`, where get/put should be quite good, but not so much delete operations (I remember looking at one implementation some time ago, and it had so many checks in place :eyes: ).

So _I think_ we can achieve the same by just filtering the collections and counting the total number of elements common to both sets?

e.g.

```diff
diff --git a/src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java b/src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java
index d4df1d7..c79023a 100644
--- a/src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java
+++ b/src/main/java/org/apache/commons/text/similarity/SorensenDicesSimilarity.java
@@ -18,6 +18,7 @@ package org.apache.commons.text.similarity;
 
 import java.util.HashSet;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
  * A similarity algorithm indicating the percentage of matched characters
@@ -93,8 +94,9 @@ public class SorensenDicesSimilarity implements SimilarityScore<Double> {
         Set<String> nRight = createBigrams(right);
 
         final int total = nLeft.size() + nRight.size();
-        nLeft.retainAll(nRight);
-        final int intersection = nLeft.size();
+        final long intersection = nLeft.stream()
+                .filter(nRight::contains)
+                .collect(Collectors.counting());
 
         return (2.0d * intersection) / total;
     }
```

I believe this way we take advantage of Java 8 syntax, and avoid creating a new data structure, or performing multiple checks. The tests worked, but being quite late Sunday for me, almost time to call it a day, and my last comment in this pull request, I may be saying something totally silly. So no hard feelings if anyone corrects me :+1:

Reply:
Nice. I had suggested `retailAll` as it is used in `CosineSimilarity`.

Reply:
If this sounds like a good approach, I can create a JIRA for updating that one too later (thanks for the pointer).